src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -249,16 +246,30 @@ public void leaveToken(DetailAST ast) {
     /**
      * Checks whether a class should be declared as final or not.
      *
-     * @param desc description of the class
+     * @param classDesc description of the class
      * @return true if given class should be declared as final otherwise false
      */
-    private static boolean shouldBeDeclaredAsFinal(ClassDesc desc) {
-        return desc.isWithPrivateCtor()
-                && !(desc.isDeclaredAsAbstract()
-                    || desc.isSuperClassOfAnonymousInnerClass())
-                && !desc.isDeclaredAsFinal()
-                && !desc.isWithNonPrivateCtor()
-                && !desc.isWithNestedSubclass();
+    private static boolean shouldBeDeclaredAsFinal(ClassDesc classDesc) {
+        final boolean isAbstract = classDesc.isDeclaredAsAbstract();
+        final boolean isSuperClassOfAnonymousInnerClass =
+                classDesc.isSuperClassOfAnonymousInnerClass();
+        final boolean declaredAsFinal = classDesc.isDeclaredAsFinal();
+        final boolean withNestedSubClass = classDesc.isWithNestedSubclass();
```

Comment:
Remove such variables , it should be clear that we do logic on classDesc

Reply:
done

logic not changed so regression is not required